front/pages/api/w/[wId]/vaults/[vId]/data_sources/index.ts

```diff
@@ -0,0 +1,637 @@
+      embedder_config: {
+        embedder: {
+          max_chunk_size: embedderConfig.max_chunk_size,
+          model_id: embedderConfig.model_id,
+          provider_id: embedderConfig.provider_id,
+          splitter_id: embedderConfig.splitter_id,
+        },
+      },
+      qdrant_config: {
+        cluster: DEFAULT_QDRANT_CLUSTER,
+        shadow_write_cluster: null,
+      },
+    },
+    credentials: dustManagedCredentials(),
+  });
+
+  if (dustDataSource.isErr()) {
+    return apiError(req, res, {
+      status_code: 500,
+      api_error: {
+        type: "internal_server_error",
+        message: "Failed to create the data source.",
+        data_source_error: dustDataSource.error,
+      },
+    });
+  }
+
+  const dataSource = await DataSourceResource.makeNew(
+    {
+      assistantDefaultSelected:
+        isConnectorProviderAssistantDefaultSelected(provider),
+      connectorProvider: provider,
+      description: dataSourceDescription,
+      dustAPIProjectId: dustProject.value.project.project_id.toString(),
+      editedByUserId: user.id,
+      name: dataSourceName,
+      workspaceId: owner.id,
+    },
+    vault
+  );
+
+  // For each data source, we create two views:
+  // - One default view in its associated vault
+  // - If the data source resides in the system vault, we also create a custom view in the global vault until vault are released.
+  if (dataSource.vault.isSystem()) {
+    const globalVault = await VaultResource.fetchWorkspaceGlobalVault(auth);
+
+    await DataSourceViewResource.createViewInVaultFromDataSourceIncludingAllDocuments(
+      globalVault,
+      dataSource,
+      "custom"
+    );
+  }
+
+  const dataSourceView =
+    await DataSourceViewResource.createViewInVaultFromDataSourceIncludingAllDocuments(
+      dataSource.vault,
+      dataSource
+    );
+
+  const connectorsAPI = new ConnectorsAPI(
+    config.getConnectorsAPIConfig(),
+    logger
+  );
+
+  const connectorsRes = await connectorsAPI.createConnector(
+    provider,
+    owner.sId,
+    systemAPIKeyRes.value.secret,
+    dataSourceName,
+    connectionId ?? "none",
+    configuration
+  );
+
+  if (connectorsRes.isErr()) {
+    logger.error(
+      {
+        error: connectorsRes.error,
+      },
+      "Failed to create the connector"
+    );
+    await dataSource.delete(auth);
+    const deleteRes = await coreAPI.deleteDataSource({
+      projectId: dustProject.value.project.project_id.toString(),
+      dataSourceName: dustDataSource.value.data_source.data_source_id,
+    });
+    if (deleteRes.isErr()) {
+      logger.error(
+        {
+          error: deleteRes.error,
+        },
+        "Failed to delete the data source"
+      );
+    }
+    return apiError(req, res, {
+      status_code: 500,
+      api_error: {
+        type: "internal_server_error",
+        message: "Failed to create the connector.",
+        connectors_error: connectorsRes.error,
+      },
+    });
+  }
+  const email = auth.user()?.email;
```

Comment:
this should happen only after everything is complete not before

Reply:
Should happen after we sent the response?

Reply:
I moved it just before we send the response, since it's void it should be non blocking sending the response.